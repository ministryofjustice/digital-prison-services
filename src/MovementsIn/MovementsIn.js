import React from 'react'
import PropTypes from 'prop-types'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'
import SortableColumn from '../tablesorting/SortableColumn'
import OffenderName from '../OffenderName'
import DateFormatter from '../DateFormatter'
import OffenderLink from '../OffenderLink'
import OffenderImage from '../OffenderImage'
import SortLov from '../tablesorting/SortLov'
import Location from '../Location'
import HoursAndMinutes from '../HoursAndMinutes'
import { LAST_NAME } from '../tablesorting/sortColumns'
import Flags from '../FullFlags/Flags'

const shouldBeClickable = (row, agencyId) => Boolean(row && row.toAgencyId === agencyId && row.location)

const withLink = (row, activeCaseLoadId, element) => {
  if (shouldBeClickable(row, activeCaseLoadId)) {
    return <OffenderLink offenderNo={row.offenderNo}>{element}</OffenderLink>
  }
  return element
}

const MovementsIn = ({ rows, sortOrder, setColumnSort, agencyId }) => (
  <>
    <GridRow>
      <GridCol setWidth="one-quarter">
        <SortLov sortColumns={[LAST_NAME]} sortColumn={LAST_NAME} sortOrder={sortOrder} setColumnSort={setColumnSort} />
      </GridCol>
    </GridRow>
    <table>
      <thead>
        <tr>
          {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
          <th className="straight width10" />
          <th className="straight width10">
            <SortableColumn
              heading="Name"
              column={LAST_NAME}
              sortOrder={sortOrder}
              setColumnSort={setColumnSort}
              sortColumn={LAST_NAME}
            />
          </th>
          <th className="straight width10">Prison no.</th>
          <th className="straight width10">D.O.B.</th>
          <th className="straight width10">Incentive Level</th>
          <th className="straight width10">Location</th>
          <th className="straight width10">Time in</th>
          <th className="straight width15">Arrived from</th>
          <th className="straight width15">Flags</th>
        </tr>
      </thead>
      <tbody>
        {rows.map(row => (
          <tr
            className="row-gutters"
            key={[row.offenderNo, row.fromAgencyDescription, row.movementsOut, row.location, row.toAgencyId].join('_')}
          >
            <td className="row-gutters">{withLink(row, agencyId, <OffenderImage offenderNo={row.offenderNo} />)}</td>
            <td className="row-gutters">
              {withLink(row, agencyId, <OffenderName lastName={row.lastName} firstName={row.firstName} />)}
            </td>
            <td className="row-gutters">{row.offenderNo}</td>
            <td className="row-gutters">
              <DateFormatter isoDate={row.dateOfBirth} />
            </td>
            <td>{row.iepLevel}</td>
            <td className="row-gutters">
              <Location location={row.location} />
            </td>
            <td>
              <HoursAndMinutes hhmmss={row.movementTime} />
            </td>
            <td>{row.fromAgencyDescription || row.fromCity}</td>
            <td className="row-gutters">
              <Flags offenderNo={row.offenderNo} alerts={row.alerts} category={row.category} />
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  </>
)

MovementsIn.propTypes = {
  rows: PropTypes.arrayOf(
    PropTypes.shape({
      offenderNo: PropTypes.string.isRequired,
      firstName: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
      location: PropTypes.string,
      movementTime: PropTypes.string,
      fromAgencyDescription: PropTypes.string,
      alerts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
      category: PropTypes.string,
      fromCity: PropTypes.string,
    })
  ),
  sortOrder: PropTypes.string,
  setColumnSort: PropTypes.func,
  agencyId: PropTypes.string,
}
MovementsIn.defaultProps = {
  rows: [],
  sortOrder: '',
  setColumnSort: () => {},
  agencyId: '',
}

export default MovementsIn
