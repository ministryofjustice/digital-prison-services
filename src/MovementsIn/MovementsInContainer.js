import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import axios from 'axios'

import MovementsIn from './MovementsIn'
import { resetError, setLoaded } from '../redux/actions'
import { ASC } from '../tablesorting/sortOrder'
import Page from '../Components/Page'
import routePaths from '../routePaths'

import SortableDataSource from '../tablesorting/SortableDataSource'
import { lastNameComparator } from '../tablesorting/comparatorComposition'

class MovementsInContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      movementsIn: [],
    }
  }

  componentDidMount() {
    this.getMovementsIn()
  }

  componentDidUpdate(previousProps) {
    const { agencyId, history } = this.props

    if (agencyId !== previousProps.agencyId) {
      history.replace(routePaths.establishmentRoll)
    }
  }

  async getMovementsIn() {
    const { agencyId, handleError, resetErrorDispatch, setLoadedDispatch } = this.props
    try {
      resetErrorDispatch()
      setLoadedDispatch(false)
      const response = await axios.get(`/api/movements/${agencyId}/in`)
      const movementsIn = response.data
      this.setState({ movementsIn })
      setLoadedDispatch(true)
    } catch (error) {
      handleError(error)
    }
  }

  render() {
    const { movementsIn } = this.state
    const { agencyId } = this.props

    return (
      <Page title="In today">
        <SortableDataSource sortOrder={ASC} rows={movementsIn} comparator={lastNameComparator} agencyId={agencyId}>
          <MovementsIn />
        </SortableDataSource>
      </Page>
    )
  }
}

MovementsInContainer.propTypes = {
  handleError: PropTypes.func.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func.isRequired,
  }).isRequired,

  agencyId: PropTypes.string.isRequired,
  resetErrorDispatch: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  agencyId: state.app.user.activeCaseLoadId,
})

const mapDispatchToProps = dispatch => ({
  resetErrorDispatch: () => dispatch(resetError()),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MovementsInContainer)
