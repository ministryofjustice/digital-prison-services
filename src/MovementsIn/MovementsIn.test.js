import React from 'react'
import { shallow } from 'enzyme'

import MovementsIn from './MovementsIn'

describe('Movements in', () => {
  it('should render page correctly', () => {
    const movements = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        location: 'MDI-1-3-017',
        movementTime: '12:01:02',
        fromAgencyDescription: 'Hull (HMP)',
        toAgencyId: 'MDI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={movements} agencyId="MDI" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should remove the offender link when the location is not set', () => {
    const movements = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        movementTime: '12:01:02',
        fromAgencyDescription: 'Hull (HMP)',
        toAgencyCode: 'HLI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={movements} agencyId="HLI" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should remove the offender link when the toAgencyId does not match the current agencyId/active case load', () => {
    const movements = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        movementTime: '12:01:02',
        location: 'CELL-1',
        fromAgencyDescription: 'Hull (HMP)',
        toAgencyCode: 'HLI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={movements} agencyId="LEI" />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should display fromCity when fromAgencyDescription is not set ', () => {
    const offenders = [
      {
        offenderNo: 'G0000GG',
        firstName: 'firstName',
        lastName: 'lastName',
        movementTime: '12:01:02',
        location: 'CELL-1',
        fromCity: 'York',
        toAgencyCode: 'HLI',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsIn sortOrder="ASC" setColumnSort={jest.fn()} rows={offenders} agencyId="LEI" />)
    expect(wrapper).toMatchSnapshot()
  })
})
