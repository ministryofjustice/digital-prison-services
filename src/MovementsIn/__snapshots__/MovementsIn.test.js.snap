// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movements in should display fromCity when fromAgencyDescription is not set  1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-quarter"
    >
      <SortLov
        setColumnSort={[MockFunction]}
        sortColumn="lastName"
        sortColumns={
          Array [
            "lastName",
          ]
        }
        sortOrder="ASC"
      />
    </GridCol>
  </GridRow>
  <table>
    <thead>
      <tr>
        <th
          className="straight width10"
        />
        <th
          className="straight width10"
        >
          <SortableColumn
            column="lastName"
            heading="Name"
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortOrder="ASC"
          />
        </th>
        <th
          className="straight width10"
        >
          Prison no.
        </th>
        <th
          className="straight width10"
        >
          D.O.B.
        </th>
        <th
          className="straight width10"
        >
          Incentive Level
        </th>
        <th
          className="straight width10"
        >
          Location
        </th>
        <th
          className="straight width10"
        >
          Time in
        </th>
        <th
          className="straight width15"
        >
          Arrived from
        </th>
        <th
          className="straight width15"
        >
          Flags
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
        key="G0000GG___CELL-1_"
      >
        <td
          className="row-gutters"
        >
          <OffenderImage
            offenderNo="G0000GG"
          />
        </td>
        <td
          className="row-gutters"
        >
          <OffenderName
            firstName="firstName"
            lastName="lastName"
          />
        </td>
        <td
          className="row-gutters"
        >
          G0000GG
        </td>
        <td
          className="row-gutters"
        >
          <FormattedDate />
        </td>
        <td />
        <td
          className="row-gutters"
        >
          <Connect(Location)
            location="CELL-1"
          />
        </td>
        <td>
          <HoursAndMinutes
            hhmmss="12:01:02"
          />
        </td>
        <td>
          York
        </td>
        <td
          className="row-gutters"
        >
          <Flags
            alerts={
              Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ]
            }
            category=""
            offenderNo="G0000GG"
          />
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`Movements in should remove the offender link when the location is not set 1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-quarter"
    >
      <SortLov
        setColumnSort={[MockFunction]}
        sortColumn="lastName"
        sortColumns={
          Array [
            "lastName",
          ]
        }
        sortOrder="ASC"
      />
    </GridCol>
  </GridRow>
  <table>
    <thead>
      <tr>
        <th
          className="straight width10"
        />
        <th
          className="straight width10"
        >
          <SortableColumn
            column="lastName"
            heading="Name"
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortOrder="ASC"
          />
        </th>
        <th
          className="straight width10"
        >
          Prison no.
        </th>
        <th
          className="straight width10"
        >
          D.O.B.
        </th>
        <th
          className="straight width10"
        >
          Incentive Level
        </th>
        <th
          className="straight width10"
        >
          Location
        </th>
        <th
          className="straight width10"
        >
          Time in
        </th>
        <th
          className="straight width15"
        >
          Arrived from
        </th>
        <th
          className="straight width15"
        >
          Flags
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
        key="G0000GG_Hull (HMP)___"
      >
        <td
          className="row-gutters"
        >
          <OffenderImage
            offenderNo="G0000GG"
          />
        </td>
        <td
          className="row-gutters"
        >
          <OffenderName
            firstName="firstName"
            lastName="lastName"
          />
        </td>
        <td
          className="row-gutters"
        >
          G0000GG
        </td>
        <td
          className="row-gutters"
        >
          <FormattedDate />
        </td>
        <td />
        <td
          className="row-gutters"
        >
          <Connect(Location) />
        </td>
        <td>
          <HoursAndMinutes
            hhmmss="12:01:02"
          />
        </td>
        <td>
          Hull (HMP)
        </td>
        <td
          className="row-gutters"
        >
          <Flags
            alerts={
              Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ]
            }
            category=""
            offenderNo="G0000GG"
          />
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`Movements in should remove the offender link when the toAgencyId does not match the current agencyId/active case load 1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-quarter"
    >
      <SortLov
        setColumnSort={[MockFunction]}
        sortColumn="lastName"
        sortColumns={
          Array [
            "lastName",
          ]
        }
        sortOrder="ASC"
      />
    </GridCol>
  </GridRow>
  <table>
    <thead>
      <tr>
        <th
          className="straight width10"
        />
        <th
          className="straight width10"
        >
          <SortableColumn
            column="lastName"
            heading="Name"
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortOrder="ASC"
          />
        </th>
        <th
          className="straight width10"
        >
          Prison no.
        </th>
        <th
          className="straight width10"
        >
          D.O.B.
        </th>
        <th
          className="straight width10"
        >
          Incentive Level
        </th>
        <th
          className="straight width10"
        >
          Location
        </th>
        <th
          className="straight width10"
        >
          Time in
        </th>
        <th
          className="straight width15"
        >
          Arrived from
        </th>
        <th
          className="straight width15"
        >
          Flags
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
        key="G0000GG_Hull (HMP)__CELL-1_"
      >
        <td
          className="row-gutters"
        >
          <OffenderImage
            offenderNo="G0000GG"
          />
        </td>
        <td
          className="row-gutters"
        >
          <OffenderName
            firstName="firstName"
            lastName="lastName"
          />
        </td>
        <td
          className="row-gutters"
        >
          G0000GG
        </td>
        <td
          className="row-gutters"
        >
          <FormattedDate />
        </td>
        <td />
        <td
          className="row-gutters"
        >
          <Connect(Location)
            location="CELL-1"
          />
        </td>
        <td>
          <HoursAndMinutes
            hhmmss="12:01:02"
          />
        </td>
        <td>
          Hull (HMP)
        </td>
        <td
          className="row-gutters"
        >
          <Flags
            alerts={
              Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ]
            }
            category=""
            offenderNo="G0000GG"
          />
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`Movements in should render page correctly 1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-quarter"
    >
      <SortLov
        setColumnSort={[MockFunction]}
        sortColumn="lastName"
        sortColumns={
          Array [
            "lastName",
          ]
        }
        sortOrder="ASC"
      />
    </GridCol>
  </GridRow>
  <table>
    <thead>
      <tr>
        <th
          className="straight width10"
        />
        <th
          className="straight width10"
        >
          <SortableColumn
            column="lastName"
            heading="Name"
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortOrder="ASC"
          />
        </th>
        <th
          className="straight width10"
        >
          Prison no.
        </th>
        <th
          className="straight width10"
        >
          D.O.B.
        </th>
        <th
          className="straight width10"
        >
          Incentive Level
        </th>
        <th
          className="straight width10"
        >
          Location
        </th>
        <th
          className="straight width10"
        >
          Time in
        </th>
        <th
          className="straight width15"
        >
          Arrived from
        </th>
        <th
          className="straight width15"
        >
          Flags
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
        key="G0000GG_Hull (HMP)__MDI-1-3-017_MDI"
      >
        <td
          className="row-gutters"
        >
          <OffenderLink
            alerts={false}
            offenderNo="G0000GG"
          >
            <OffenderImage
              offenderNo="G0000GG"
            />
          </OffenderLink>
        </td>
        <td
          className="row-gutters"
        >
          <OffenderLink
            alerts={false}
            offenderNo="G0000GG"
          >
            <OffenderName
              firstName="firstName"
              lastName="lastName"
            />
          </OffenderLink>
        </td>
        <td
          className="row-gutters"
        >
          G0000GG
        </td>
        <td
          className="row-gutters"
        >
          <FormattedDate />
        </td>
        <td />
        <td
          className="row-gutters"
        >
          <Connect(Location)
            location="MDI-1-3-017"
          />
        </td>
        <td>
          <HoursAndMinutes
            hhmmss="12:01:02"
          />
        </td>
        <td>
          Hull (HMP)
        </td>
        <td
          className="row-gutters"
        >
          <Flags
            alerts={
              Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ]
            }
            category=""
            offenderNo="G0000GG"
          />
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;
